rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,512 @@
+---
+authors: Trent Clarke (trent@goteleport.com)
+state: draft
+---
+
+# RFD 2 - Custom Terraform Registry
+
+## What
+
+A Terraform _registry_ is an index of available terraform _providers_. The
+Terraform tool uses registries to locate, validate and install Terraform
+providers for the various services required by terraform scripts. Hashicorp
+(the Terraform vendor) publishes the protocol that Terraform uses to interact
+with registries and so third parties (like us) may provide _custom_ Terraform
+registries outside the main, Hashicorp-provided public registry.
+
+This RFD investigates the use of a custom Terraform registry to distribute the
+Teleport provider for Terraform.
+
+## Why
+
+We have two main goals that effect how we offer our Terraform provider to
+customers
+
+ 1. We want to meet out customers where they are, e.g. if they're using
+    Terraform, we want to supply them with a Terraform provider, and a method
+    for installing it that is natural to the Terraform ecosystem. This is a
+    good practice in general and, in the specific, is part of our Q1 focus on
+    improving our Time-to_first-Value.
+
+ 2. We want to maintain close control of the distribution of our software,
+    including supply chain management and monitoring of publishing, changes,
+    downloads and host of other activities.
```

Comment:
This doesn't really answer _why_ - it's an answer, of course, but it doesn't actually offer much detail explaining why this solution is explicitly better than the public Terraform registry. Could you maybe add a few examples of situations where this would benefit us? Explain the limitations of the public Terraform registry?

Reply:
Biggest issue I can see is that they require OAuth access to our GitHub organization, and want write access to manage webhooks in all our repositories.

Reply:
We can create a separate GitHub organization for this purpose easily enough.

I'll reach out to HashiCorp about this to see why they need such permissions.

Reply:
It seems like a vast overreach in permission terms as well.